electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +263,107 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
     if pwr_lo is not None:
       cast(Block, builder.get_enclosing_block()).connect(pwr_lo, self.pwr_lo)
     return self
+
+
+class SoftPowerGate(PowerSwitch, KiCadSchematicBlock, Block):  # migrate from the multimater
+  """A high-side PFET power gate that has a button to power on, can be latched
+  on by an external signal, and provides the button output as a signal.
+
+  Attributes:
+    pull_resistance: The resistance of the pull-up resistor.
+    amp_resistance: The resistance of the amplifier resistor.
+    diode_drop: The voltage drop across the diode.
+
+  Ports:
+    pwr_in: A voltage sink port for the input power.
+    pwr_out: A voltage source port for the output power.
+    gnd: ground.
+    btn_out: A digital single source port for the button output signal.
+    btn_in: A digital single source port for the button input signal. Should be connected to a button. Do not connect IO.
+    control: A digital sink port for external control to latch the power on.
+  """
+  @init_in_parent
+  def __init__(self, pull_resistance: RangeLike=10*kOhm(tol=0.05), amp_resistance: RangeLike=10*kOhm(tol=0.05), diode_drop:RangeLike=(0, 0.4)*Volt):
+    super().__init__()
+    self.pwr_in = self.Port(VoltageSink.empty(), [Input])
+    self.pwr_out = self.Port(VoltageSource.empty(), [Output])
+    self.gnd = self.Port(Ground.empty(), [Common])
+
+    self.btn_out = self.Port(DigitalSingleSource.empty())
+    self.btn_in = self.Port(DigitalSingleSource.empty())
+    self.control = self.Port(DigitalSink.empty())  # digital level control - gnd-referenced NFET gate
+
+    self.pull_resistance = self.ArgParameter(pull_resistance)
+    self.amp_resistance = self.ArgParameter(amp_resistance)
+    self.diode_drop = self.ArgParameter(diode_drop)
+
+  def contents(self):
+    super().contents()
+
+    control_voltage = self.control.link().voltage.hull(RangeExpr.ZERO,)
```

Comment:
If this is the max voltage across ctl_diode and btn_diode, it may be more correct to use the voltage on the btn_in port

Reply:
>I'm on the fence as to whether a convenience wrapper for the SoftPowerGate + switch makes sense, since it does so little extra. But on the positive side, it's more error-proof in the common case, while advanced users can use custom switches like encoders - and I think it'd be worth it? Thoughts?

I think it is good to add SoftPowerGate + Switch  If people anyway need to add a button at least always.
 In most cases, people would do refinement on the switch to swap up the button and it should be enough I assume?
 
 > switches like encoders
 Spin to turn on 🙃

Reply:
Composition vs Inheritance and use composition for this

Reply:
'    control_voltage = self.control.link().voltage.hull(ground)'